nighthawk/source/server/http_test_server_filter.cc

```diff
@@ -0,0 +1,83 @@
+#include <string>
+
+#include "http_test_server_filter.h"
+
+#include "absl/strings/numbers.h"
+
+#include "common/protobuf/utility.h"
+#include "envoy/server/filter_config.h"
+
+namespace Nighthawk {
+namespace Server {
+
+HttpTestServerDecoderFilterConfig::HttpTestServerDecoderFilterConfig(
+    const nighthawk::server::TestServer& proto_config)
+    : server_config_(proto_config) {}
+
+HttpTestServerDecoderFilter::HttpTestServerDecoderFilter(
+    HttpTestServerDecoderFilterConfigSharedPtr config)
+    : config_(config) {}
+
+HttpTestServerDecoderFilter::~HttpTestServerDecoderFilter() {}
+
+void HttpTestServerDecoderFilter::onDestroy() {}
+
+Envoy::Http::FilterHeadersStatus
+HttpTestServerDecoderFilter::decodeHeaders(Envoy::Http::HeaderMap& headers, bool) {
+  const auto request_config_header = headers.get(TestServer::HeaderNames::get().TestServerConfig);
+  nighthawk::server::TestServer base_config = config_->server_config();
+
+  bool error = false;
+  if (request_config_header != nullptr) {
+    try {
+      nighthawk::server::TestServer json_config;
+      Envoy::MessageUtil::loadFromJson(request_config_header->value().c_str(), json_config);
+      base_config.MergeFrom(json_config);
+    } catch (Envoy::EnvoyException) {
+      error = true;
+    }
+  }
+
+  // TODO(oschaaf): figure out how to get _cc_validate to be around in BUILD, and
+  // do something like this. Enable relevant tests when done.
+  // error = base_config.Validate();
+
+  if (error) {
+    decoder_callbacks_->sendLocalReply(static_cast<Envoy::Http::Code>(500),
+                                       "test-server didn't understand the request", nullptr,
```

Comment:
It might be helpful to include the error message inside the `EnvoyException` above in the response body.